src/psyclone/nemo.py

```diff
@@ -377,14 +388,18 @@ class NemoKern(CodedKern):
 
     '''
     def __init__(self, psyir_nodes, parse_tree, parent=None):
-        # pylint: disable=super-init-not-called
+        super(NemoKern, self).__init__()
         from psyclone.psyGen import KernelSchedule
         self._name = ""
         self._parent = parent
         # The corresponding set of nodes in the fparser2 parse tree
         self._ast = parse_tree
-        # Create a kernel schedule
-        self._kern_schedule = KernelSchedule(self._name)
+        # Create a kernel schedule. Since in NEMO all functions are
+        # inlined, set this NemoKern as parent of the KernelSchedule.
+        # This will allow the dependency analysis going up from
+        # a kernel to the loops (which is required for declaring
+        # private variables).
+        self._kern_schedule = KernelSchedule(self._name, self)
```

Comment:
I hope I have done what you wanted, but unfortunately this caused a bit of a problem. In transformations.py is the code to check if the node types are valid:
```
                flat_list = [item for item in child.walk(object)
                             if not isinstance(item, Schedule)]
```
Now this loop returns all PSyIR nodes (e.g. assignments, array references, ...) - and then declares that the transformation can not be applied. As an immediate fix I have just disabled this test (*grin* glad that the option-for-transforms pull request was merged). But that's obviously not  a good idea.

My suggested solution:
We add a new class 'Statement(Node)', which is otherwise empty. Then all statement nodes (IfBlock, Loop, ...) will be based on Statement (instead of node). Then the walk statement can collect only Statement nodes.

Alternative: we add a flag 'is_statement'?? Suggestions welcome :)

Reply:
I don't quite understand. Are we talking about line 164 in transformations.py? It seems to me that it's doing what we want. Are you worried about the performance implications? Perhaps it would be better if we had `self.invalid_node_types` instead of `valid_node_types`? Maybe we need to have a chat about this?